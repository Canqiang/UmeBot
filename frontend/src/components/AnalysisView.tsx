import React from 'react';
import { TrendingUp, TrendingDown, AlertCircle, CheckCircle, Target, Zap } from 'lucide-react';

interface AnalysisViewProps {
  data: any;
}

export const AnalysisView: React.FC<AnalysisViewProps> = ({ data }) => {
  // 因果效应数据
  const causalEffects = data.causal_effects || [];
  const interactions = data.interactions || [];
  const recommendations = data.recommendations || [];
  const forecast = data.forecast || null;

  return (
    <div className="space-y-6">
      {/* 因果效应分析 */}
      {causalEffects.length > 0 && (
        <div className="bg-white rounded-xl border border-gray-200 p-6">
          <div className="flex items-center mb-4">
            <Zap className="w-5 h-5 text-purple-500 mr-2" />
            <h3 className="text-lg font-semibold text-gray-800">因果效应分析</h3>
          </div>

          <div className="space-y-4">
            {causalEffects.map((effect: any, index: number) => (
              <div key={index} className="bg-gray-50 rounded-lg p-4">
                <div className="flex items-center justify-between mb-2">
                  <span className="font-medium text-gray-700">{effect.factor}</span>
                  <span className={`text-sm font-semibold ${
                    effect.significant ? 'text-green-600' : 'text-gray-500'
                  }`}>
                    {effect.significant ? '显著' : '不显著'}
                  </span>
                </div>

                <div className="flex items-center space-x-4 text-sm">
                  <div>
                    <span className="text-gray-500">效应值: </span>
                    <span className="font-medium">
                      {effect.effect > 0 ? '+' : ''}{effect.effect.toFixed(2)}%
                    </span>
                  </div>
                  <div>
                    <span className="text-gray-500">置信区间: </span>
                    <span className="font-medium">
                      [{effect.confidence_interval[0].toFixed(2)}, {effect.confidence_interval[1].toFixed(2)}]
                    </span>
                  </div>
                  {effect.sample_size && (
                    <div>
                      <span className="text-gray-500">样本量: </span>
                      <span className="font-medium">{effect.sample_size.toLocaleString()}</span>
                    </div>
                  )}
                </div>

                {/* 效应值可视化 */}
                <div className="mt-3">
                  <div className="w-full bg-gray-200 rounded-full h-2">
                    <div
                      className={`h-2 rounded-full ${
                        effect.effect > 0 ? 'bg-green-500' : 'bg-red-500'
                      }`}
                      style={{ width: `${Math.min(Math.abs(effect.effect), 100)}%` }}
                    />
                  </div>
                </div>
              </div>
            ))}
          </div>
        </div>
      )}

      {/* 交互效应 */}
      {interactions.length > 0 && (
        <div className="bg-white rounded-xl border border-gray-200 p-6">
          <div className="flex items-center mb-4">
            <Target className="w-5 h-5 text-blue-500 mr-2" />
            <h3 className="text-lg font-semibold text-gray-800">交互效应</h3>
          </div>

          <div className="grid grid-cols-1 md:grid-cols-2 gap-4">
            {interactions.map((interaction: any, index: number) => (
              <div key={index} className="bg-blue-50 rounded-lg p-4">
                <div className="font-medium text-gray-700 mb-2">
                  {interaction.factors.join(' × ')}
                </div>
                <div className="text-2xl font-bold text-blue-600">
                  {interaction.interaction_effect > 0 ? '+' : ''}
                  {interaction.interaction_effect.toFixed(1)}%
                </div>
                {interaction.combined_effect && (
                  <div className="text-sm text-gray-500 mt-1">
                    综合效应: {interaction.combined_effect.toFixed(1)}%
                  </div>
                )}
              </div>
            ))}
          </div>
        </div>
      )}

      {/* 预测结果 */}
      {forecast && (
        <div className="bg-white rounded-xl border border-gray-200 p-6">
          <div className="flex items-center mb-4">
            <TrendingUp className="w-5 h-5 text-green-500 mr-2" />
            <h3 className="text-lg font-semibold text-gray-800">销售预测</h3>
          </div>

          <div className="grid grid-cols-2 md:grid-cols-4 gap-4">
            <div className="text-center">
              <div className="text-sm text-gray-500 mb-1">预测总额</div>
              <div className="text-xl font-bold text-gray-800">
                ${forecast.total_forecast?.toLocaleString()}
              </div>
            </div>
            <div className="text-center">
              <div className="text-sm text-gray-500 mb-1">日均预测</div>
              <div className="text-xl font-bold text-gray-800">
                ${forecast.avg_daily_forecast?.toFixed(0)}
              </div>
            </div>
            <div className="text-center">
              <div className="text-sm text-gray-500 mb-1">最高预测</div>
              <div className="text-xl font-bold text-green-600">
                ${forecast.max_daily_forecast?.toFixed(0)}
              </div>
            </div>
            <div className="text-center">
              <div className="text-sm text-gray-500 mb-1">最低预测</div>
              <div className="text-xl font-bold text-red-600">
                ${forecast.min_daily_forecast?.toFixed(0)}
              </div>
            </div>
          </div>
        </div>
      )}

      {/* 建议和行动方案 */}
      {recommendations.length > 0 && (
        <div className="bg-gradient-to-r from-green-50 to-emerald-50 rounded-xl p-6">
          <div className="flex items-center mb-4">
            <CheckCircle className="w-5 h-5 text-green-600 mr-2" />
            <h3 className="text-lg font-semibold text-gray-800">行动建议</h3>
          </div>

          <div className="space-y-3">
            {recommendations.map((rec: string, index: number) => (
              <div key={index} className="flex items-start">
                <span className="flex-shrink-0 w-6 h-6 bg-green-500 text-white rounded-full flex items-center justify-center text-xs font-bold mr-3">
                  {index + 1}
                </span>
                <p className="text-gray-700 flex-1">{rec}</p>
              </div>
            ))}
          </div>
        </div>
      )}

      {/* 警告信息 */}
      {data.warnings && data.warnings.length > 0 && (
        <div className="bg-yellow-50 border border-yellow-200 rounded-xl p-4">
          <div className="flex items-center mb-2">
            <AlertCircle className="w-5 h-5 text-yellow-600 mr-2" />
            <h4 className="font-semibold text-yellow-800">注意事项</h4>
          </div>
          <ul className="space-y-1 text-sm text-yellow-700">
            {data.warnings.map((warning: string, index: number) => (
              <li key={index}>• {warning}</li>
            ))}
          </ul>
        </div>
      )}
    </div>
  );
};