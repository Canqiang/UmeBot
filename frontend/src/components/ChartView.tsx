import React, { useEffect } from 'react';
import { BarChart3 } from 'lucide-react';
import * as echarts from 'echarts';

interface ChartViewProps {
  data: {
    type: 'line' | 'bar' | 'pie' | 'scatter' | 'mixed';
    title?: string;
    series: any[];
    xAxis?: any;
    yAxis?: any;
    options?: any;
  };
  onPointClick?: (params: any) => void;
}

export const ChartView: React.FC<ChartViewProps> = ({ data, onPointClick }) => {
  const chartRef = React.useRef<HTMLDivElement>(null);

  useEffect(() => {
    if (chartRef.current) {
      const chart = echarts.init(chartRef.current);

      const option = buildChartOption(data);
      chart.setOption(option);

      const handleResize = () => chart.resize();
      const handleClick = (params: any) => {
        if (onPointClick) {
          onPointClick(params);
        }
      };

      window.addEventListener('resize', handleResize);
      chart.on('click', handleClick);

      return () => {
        window.removeEventListener('resize', handleResize);
        chart.off('click', handleClick);
        chart.dispose();
      };
    }
  }, [data, onPointClick]);

  const buildChartOption = (data: ChartViewProps['data']) => {
    const baseOption = {
      title: {
        text: data.title || '',
        left: 'center',
        top: 0
      },
      tooltip: {
        trigger: data.type === 'pie' ? 'item' : 'axis',
        backgroundColor: 'rgba(255, 255, 255, 0.95)',
        borderColor: '#e5e7eb',
        borderWidth: 1,
        textStyle: {
          color: '#374151'
        }
      },
      legend: {
        bottom: 0,
        data: data.series.map(s => s.name)
      },
      grid: {
        left: '3%',
        right: '4%',
        bottom: '15%',
        top: '10%',
        containLabel: true
      }
    };

    if (data.type === 'pie') {
      return {
        ...baseOption,
        series: [{
          type: 'pie',
          radius: ['40%', '70%'],
          center: ['50%', '50%'],
          data: data.series,
          emphasis: {
            itemStyle: {
              shadowBlur: 10,
              shadowOffsetX: 0,
              shadowColor: 'rgba(0, 0, 0, 0.5)'
            }
          },
          label: {
            formatter: '{b}: {c} ({d}%)'
          }
        }]
      };
    }

    return {
      ...baseOption,
      xAxis: data.xAxis || { type: 'category', data: [] },
      yAxis: data.yAxis || { type: 'value' },
      series: data.series.map(s => ({
        ...s,
        type: s.type || data.type || 'line',
        smooth: s.smooth !== false,
        areaStyle: s.areaStyle || (data.type === 'line' ? { opacity: 0.1 } : undefined)
      })),
      ...data.options
    };
  };

  return (
    <div className="space-y-4">
      <div className="flex items-center justify-between">
        <div className="flex items-center space-x-2">
          <BarChart3 className="w-5 h-5 text-gray-500" />
          <span className="font-medium">数据可视化</span>
        </div>
      </div>
      <div ref={chartRef} style={{ width: '100%', height: '500px' }} />
    </div>
  );
};