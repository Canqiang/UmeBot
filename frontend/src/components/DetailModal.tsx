import React, { useState, useEffect } from 'react';
import { X, Download, Filter, Search, ChevronLeft, ChevronRight, BarChart3, LineChart, PieChart, TrendingUp, Info } from 'lucide-react';
import * as echarts from 'echarts';

interface DetailModalProps {
  isOpen: boolean;
  onClose: () => void;
  title: string;
  data: any;
  type: 'table' | 'chart' | 'analysis';
}

export const DetailModal: React.FC<DetailModalProps> = ({ isOpen, onClose, title, data, type }) => {
  if (!isOpen) return null;

  const renderContent = () => {
    switch (type) {
      case 'table':
        return <TableView data={data} />;
      case 'chart':
        return <ChartView data={data} />;
      case 'analysis':
        return <AnalysisView data={data} />;
      default:
        return <div>No data available</div>;
    }
  };

  return (
    <div className="fixed inset-0 bg-black bg-opacity-50 z-50 flex items-center justify-center p-4">
      <div className="bg-white rounded-xl shadow-2xl max-w-6xl w-full max-h-[90vh] overflow-hidden">
        <div className="flex items-center justify-between p-6 border-b">
          <h2 className="text-xl font-semibold text-gray-900">{title}</h2>
          <button
            onClick={onClose}
            className="p-2 hover:bg-gray-100 rounded-lg transition-colors"
          >
            <X className="w-5 h-5 text-gray-500" />
          </button>
        </div>
        <div className="p-6 overflow-y-auto max-h-[calc(90vh-100px)]">
          {renderContent()}
        </div>
      </div>
    </div>
  );
};