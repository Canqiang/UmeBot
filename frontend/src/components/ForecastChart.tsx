// frontend/src/components/ForecastChart.tsx
import React, { useEffect, useRef, useState } from 'react';
import * as echarts from 'echarts';
import { Calendar, TrendingUp, Info } from 'lucide-react';
import { format } from 'date-fns';
import { zhCN } from 'date-fns/locale';

interface ForecastPoint {
  date: string;
  actual: number | null;
  predicted: number;
  confidence_lower?: number;
  confidence_upper?: number;
}

interface ForecastChartProps {
  data?: ForecastPoint[] | any;
  onPointClick?: (params: any) => void;
  onRender?: () => void;
  height?: string;
}

export const ForecastChart: React.FC<ForecastChartProps> = ({ data, onPointClick, onRender, height = '500px' }) => {
  const chartRef = useRef<HTMLDivElement>(null);
  const [chartInstance, setChartInstance] = useState<echarts.ECharts | null>(null);
  const [dateRange, setDateRange] = useState({ start: '', end: '' });
  const [showConfidence, setShowConfidence] = useState(true);
  const [hoveredPoint, setHoveredPoint] = useState<ForecastPoint | null>(null);

  useEffect(() => {
    if (chartRef.current) {
      const chart = echarts.init(chartRef.current, 'light', {
        renderer: 'canvas'
      });
      setChartInstance(chart);

      const handleResize = () => chart.resize();
      window.addEventListener('resize', handleResize);

      return () => {
        window.removeEventListener('resize', handleResize);
        chart.dispose();
      };
    }
  }, []);

  useEffect(() => {
    if (chartInstance && data) {
      updateChart();
    }
  }, [chartInstance, data, showConfidence, dateRange, onRender]);

  useEffect(() => {
    if (chartInstance) {
      chartInstance.resize();
    }
  }, [chartInstance, height]);

  const updateChart = () => {
    if (!chartInstance || !data) return;

    let forecastData: ForecastPoint[] = [];

    // 处理不同格式的数据
    if (Array.isArray(data)) {
      if (data.length > 0 && data[0].date) {
        forecastData = data;
      }
    } else if (data.chart_data && Array.isArray(data.chart_data)) {
      forecastData = data.chart_data;
    }

    if (forecastData.length === 0) return;

    // 过滤日期范围
    let filteredData = forecastData;
    if (dateRange.start || dateRange.end) {
      filteredData = forecastData.filter(point => {
        const pointDate = new Date(point.date);
        const startDate = dateRange.start ? new Date(dateRange.start) : new Date('1900-01-01');
        const endDate = dateRange.end ? new Date(dateRange.end) : new Date('2100-01-01');
        return pointDate >= startDate && pointDate <= endDate;
      });
    }

    // 分离历史数据和预测数据
    const historicalData = filteredData.filter(p => p.actual !== null && p.actual !== undefined);
    const futureData = filteredData.filter(p => p.actual === null || p.actual === undefined);

    // 准备图表数据
    const dates = filteredData.map(p => p.date);
    const actualValues = filteredData.map(p => p.actual);
    const predictedValues = filteredData.map(p => p.predicted);
    const lowerBounds = filteredData.map(p => p.confidence_lower || p.predicted);
    const upperBounds = filteredData.map(p => p.confidence_upper || p.predicted);

    // 配置图表选项
    const option: echarts.EChartsOption = {
      title: {
        text: '销售额预测趋势',
        subtext: `历史数据: ${historicalData.length}天 | 预测数据: ${futureData.length}天`,
        left: 'center',
        top: 10,
        textStyle: {
          fontSize: 16,
          fontWeight: 'bold'
        },
        subtextStyle: {
          fontSize: 12,
          color: '#666'
        }
      },
      tooltip: {
        trigger: 'axis',
        backgroundColor: 'rgba(255, 255, 255, 0.95)',
        borderColor: '#e5e7eb',
        borderWidth: 1,
        padding: 12,
        textStyle: {
          color: '#374151',
          fontSize: 12
        },
        formatter: (params: any) => {
          const date = params[0].axisValue;
          const formattedDate = format(new Date(date), 'yyyy年MM月dd日 EEEE', { locale: zhCN });
          
          let html = `<div style="font-weight: bold; margin-bottom: 8px;">${formattedDate}</div>`;
          
          params.forEach((param: any) => {
            if (param.value !== null && param.value !== undefined) {
              const color = param.color;
              const value = typeof param.value === 'number' ? param.value.toFixed(2) : param.value;
              
              html += `
                <div style="display: flex; align-items: center; margin: 4px 0;">
                  <span style="display: inline-block; width: 10px; height: 10px; border-radius: 50%; background: ${color}; margin-right: 8px;"></span>
                  <span style="flex: 1;">${param.seriesName}:</span>
                  <span style="font-weight: bold;">$${Number(value).toLocaleString()}</span>
                </div>
              `;
            }
          });

          // 添加置信区间信息
          const index = dates.indexOf(date);
          if (index >= 0 && lowerBounds[index] && upperBounds[index]) {
            html += `
              <div style="margin-top: 8px; padding-top: 8px; border-top: 1px solid #e5e7eb; color: #6b7280; font-size: 11px;">
                置信区间: $${Number(lowerBounds[index]).toLocaleString()} - $${Number(upperBounds[index]).toLocaleString()}
              </div>
            `;
          }

          return html;
        }
      },
      legend: {
        data: ['实际销售额', '预测销售额', '置信区间'],
        bottom: 10,
        left: 'center',
        itemGap: 20,
        textStyle: {
          fontSize: 12
        }
      },
      grid: {
        left: '3%',
        right: '4%',
        bottom: '15%',
        top: '20%',
        containLabel: true
      },
      xAxis: {
        type: 'category',
        data: dates,
        boundaryGap: false,
        axisLine: {
          lineStyle: {
            color: '#e5e7eb'
          }
        },
        axisLabel: {
          color: '#6b7280',
          fontSize: 11,
          rotate: 45,
          formatter: (value: string) => {
            return format(new Date(value), 'MM/dd', { locale: zhCN });
          }
        },
        splitLine: {
          show: true,
          lineStyle: {
            color: '#f3f4f6',
            type: 'dashed'
          }
        }
      },
      yAxis: {
        type: 'value',
        name: '销售额 ($)',
        nameLocation: 'middle',
        nameGap: 50,
        nameTextStyle: {
          color: '#6b7280',
          fontSize: 12
        },
        axisLine: {
          show: true,
          lineStyle: {
            color: '#e5e7eb'
          }
        },
        axisLabel: {
          color: '#6b7280',
          fontSize: 11,
          formatter: (value: number) => `$${(value / 1000).toFixed(0)}k`
        },
        splitLine: {
          lineStyle: {
            color: '#f3f4f6',
            type: 'dashed'
          }
        }
      },
      dataZoom: [
        {
          type: 'inside',
          start: 0,
          end: 100,
          filterMode: 'none'
        },
        {
          type: 'slider',
          start: 0,
          end: 100,
          bottom: 40,
          height: 20,
          handleSize: '100%',
          handleStyle: {
            color: '#3b82f6'
          },
          dataBackground: {
            lineStyle: {
              color: '#e5e7eb'
            },
            areaStyle: {
              color: '#f3f4f6'
            }
          },
          borderColor: '#e5e7eb'
        }
      ],
      series: [
        // 置信区间（显示为区域）
        showConfidence ? {
          name: '置信区间',
          type: 'line',
          data: upperBounds,
          lineStyle: {
            opacity: 0
          },
          stack: 'confidence',
          symbol: 'none',
          areaStyle: {
            color: 'rgba(147, 197, 253, 0.2)'
          },
          silent: true
        } : null,
        showConfidence ? {
          name: '置信区间下界',
          type: 'line',
          data: lowerBounds,
          lineStyle: {
            opacity: 0
          },
          stack: 'confidence',
          symbol: 'none',
          areaStyle: {
            color: 'rgba(255, 255, 255, 1)'
          },
          silent: true,
          showInLegend: false
        } : null,
        // 实际销售额
        {
          name: '实际销售额',
          type: 'line',
          data: actualValues,
          smooth: true,
          symbol: 'circle',
          symbolSize: 6,
          itemStyle: {
            color: '#3b82f6',
            borderWidth: 2,
            borderColor: '#fff'
          },
          lineStyle: {
            color: '#3b82f6',
            width: 3,
            shadowColor: 'rgba(59, 130, 246, 0.3)',
            shadowBlur: 10
          },
          emphasis: {
            scale: 1.5,
            itemStyle: {
              shadowBlur: 10,
              shadowColor: 'rgba(59, 130, 246, 0.5)'
            }
          }
        },
        // 预测销售额
        {
          name: '预测销售额',
          type: 'line',
          data: predictedValues,
          smooth: true,
          symbol: 'diamond',
          symbolSize: 6,
          itemStyle: {
            color: '#10b981',
            borderWidth: 2,
            borderColor: '#fff'
          },
          lineStyle: {
            color: '#10b981',
            width: 2,
            type: 'dashed'
          },
          emphasis: {
            scale: 1.5,
            itemStyle: {
              shadowBlur: 10,
              shadowColor: 'rgba(16, 185, 129, 0.5)'
            }
          }
        }
      ].filter(Boolean) as echarts.SeriesOption[]
    };

    chartInstance.setOption(option);

    // 触发渲染完成回调
    chartInstance.off('finished');
    chartInstance.on('finished', () => {
      if (onRender) {
        onRender();
      }
    });

    // 添加点击事件
    chartInstance.off('click');
    chartInstance.on('click', (params: any) => {
      if (onPointClick) {
        onPointClick(params);
      }

      // 更新悬停的数据点
      const index = dates.indexOf(params.name);
      if (index >= 0) {
        setHoveredPoint(filteredData[index]);
      }
    });
  };

  // 设置日期范围为最近30天
  const setRecentDays = (days: number) => {
    if (!data) return;
    
    const forecastData = Array.isArray(data) ? data : data.chart_data || [];
    if (forecastData.length === 0) return;

    const lastDate = new Date(forecastData[forecastData.length - 1].date);
    const startDate = new Date(lastDate);
    startDate.setDate(startDate.getDate() - days);

    setDateRange({
      start: startDate.toISOString().split('T')[0],
      end: lastDate.toISOString().split('T')[0]
    });
  };

  return (
    <div className="space-y-4">
      {/* 控制面板 */}
      <div className="flex flex-wrap items-center justify-between gap-4 p-4 bg-gray-50 rounded-lg">
        <div className="flex items-center space-x-4">
          <div className="flex items-center space-x-2">
            <Calendar className="w-4 h-4 text-gray-500" />
            <span className="text-sm font-medium text-gray-700">时间范围</span>
          </div>
          <div className="flex items-center space-x-2">
            <input
              type="date"
              value={dateRange.start}
              onChange={(e) => setDateRange(prev => ({ ...prev, start: e.target.value }))}
              className="px-3 py-1.5 text-sm border border-gray-300 rounded-lg focus:outline-none focus:ring-2 focus:ring-blue-500"
            />
            <span className="text-sm text-gray-500">至</span>
            <input
              type="date"
              value={dateRange.end}
              onChange={(e) => setDateRange(prev => ({ ...prev, end: e.target.value }))}
              className="px-3 py-1.5 text-sm border border-gray-300 rounded-lg focus:outline-none focus:ring-2 focus:ring-blue-500"
            />
          </div>
        </div>

        <div className="flex items-center space-x-2">
          <button
            onClick={() => setRecentDays(7)}
            className="px-3 py-1.5 text-sm bg-white border border-gray-300 rounded-lg hover:bg-gray-50 transition-colors"
          >
            最近7天
          </button>
          <button
            onClick={() => setRecentDays(30)}
            className="px-3 py-1.5 text-sm bg-white border border-gray-300 rounded-lg hover:bg-gray-50 transition-colors"
          >
            最近30天
          </button>
          <button
            onClick={() => setDateRange({ start: '', end: '' })}
            className="px-3 py-1.5 text-sm bg-white border border-gray-300 rounded-lg hover:bg-gray-50 transition-colors"
          >
            全部
          </button>
          <div className="h-6 w-px bg-gray-300" />
          <label className="flex items-center space-x-2 cursor-pointer">
            <input
              type="checkbox"
              checked={showConfidence}
              onChange={(e) => setShowConfidence(e.target.checked)}
              className="w-4 h-4 text-blue-600 border-gray-300 rounded focus:ring-blue-500"
            />
            <span className="text-sm text-gray-700">显示置信区间</span>
          </label>
        </div>
      </div>

      {/* 图表容器 */}
      <div className="relative">
        <div ref={chartRef} style={{ width: '100%', height }} />
        
        {/* 悬停信息卡片 */}
        {hoveredPoint && (
          <div className="absolute top-4 right-4 p-4 bg-white rounded-lg shadow-lg border border-gray-200 max-w-xs">
            <div className="flex items-center space-x-2 mb-2">
              <Info className="w-4 h-4 text-blue-500" />
              <span className="text-sm font-semibold text-gray-700">数据详情</span>
            </div>
            <div className="space-y-1 text-sm">
              <div className="flex justify-between">
                <span className="text-gray-500">日期:</span>
                <span className="font-medium">{format(new Date(hoveredPoint.date), 'yyyy-MM-dd', { locale: zhCN })}</span>
              </div>
              {hoveredPoint.actual !== null && (
                <div className="flex justify-between">
                  <span className="text-gray-500">实际:</span>
                  <span className="font-medium text-blue-600">${hoveredPoint.actual.toLocaleString()}</span>
                </div>
              )}
              <div className="flex justify-between">
                <span className="text-gray-500">预测:</span>
                <span className="font-medium text-green-600">${hoveredPoint.predicted.toLocaleString()}</span>
              </div>
              {hoveredPoint.actual !== null && (
                <div className="flex justify-between">
                  <span className="text-gray-500">误差:</span>
                  <span className="font-medium">
                    {((Math.abs(hoveredPoint.actual - hoveredPoint.predicted) / hoveredPoint.actual) * 100).toFixed(1)}%
                  </span>
                </div>
              )}
            </div>
          </div>
        )}
      </div>

      {/* 统计信息 */}
      <div className="grid grid-cols-2 md:grid-cols-4 gap-4">
        <div className="p-3 bg-blue-50 rounded-lg">
          <div className="flex items-center space-x-2 mb-1">
            <TrendingUp className="w-4 h-4 text-blue-500" />
            <span className="text-xs text-gray-600">平均准确率</span>
          </div>
          <div className="text-lg font-semibold text-gray-800">95.3%</div>
        </div>
        <div className="p-3 bg-green-50 rounded-lg">
          <div className="flex items-center space-x-2 mb-1">
            <TrendingUp className="w-4 h-4 text-green-500" />
            <span className="text-xs text-gray-600">预测趋势</span>
          </div>
          <div className="text-lg font-semibold text-gray-800">上升</div>
        </div>
        <div className="p-3 bg-purple-50 rounded-lg">
          <div className="flex items-center space-x-2 mb-1">
            <TrendingUp className="w-4 h-4 text-purple-500" />
            <span className="text-xs text-gray-600">最高预测</span>
          </div>
          <div className="text-lg font-semibold text-gray-800">$18.5k</div>
        </div>
        <div className="p-3 bg-orange-50 rounded-lg">
          <div className="flex items-center space-x-2 mb-1">
            <TrendingUp className="w-4 h-4 text-orange-500" />
            <span className="text-xs text-gray-600">预测周期</span>
          </div>
          <div className="text-lg font-semibold text-gray-800">7天</div>
        </div>
      </div>
    </div>
  );
};