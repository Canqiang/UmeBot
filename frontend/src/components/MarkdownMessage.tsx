// frontend/src/components/MarkdownMessage.tsx
import React from 'react';
import ReactMarkdown from 'react-markdown';
import remarkGfm from 'remark-gfm';
import remarkMath from 'remark-math';
import rehypeKatex from 'rehype-katex';
import rehypeHighlight from 'rehype-highlight';
import { CheckCircle, AlertCircle, Info, AlertTriangle, Copy, Check } from 'lucide-react';

interface MarkdownMessageProps {
  content: string;
  className?: string;
}

export const MarkdownMessage: React.FC<MarkdownMessageProps> = ({ content, className = '' }) => {
  const [copiedCode, setCopiedCode] = React.useState<string | null>(null);

  const copyToClipboard = (code: string, id: string) => {
    navigator.clipboard.writeText(code);
    setCopiedCode(id);
    setTimeout(() => setCopiedCode(null), 2000);
  };

  return (
    <div className={`markdown-body ${className}`}>
      <ReactMarkdown
        remarkPlugins={[remarkGfm, remarkMath]}
        rehypePlugins={[rehypeKatex, rehypeHighlight]}
        components={{
        // 标题样式
        h1: ({ children }) => (
          <h1 className="text-2xl font-bold mt-6 mb-4 text-gray-900">{children}</h1>
        ),
        h2: ({ children }) => (
          <h2 className="text-xl font-semibold mt-5 mb-3 text-gray-800">{children}</h2>
        ),
        h3: ({ children }) => (
          <h3 className="text-lg font-medium mt-4 mb-2 text-gray-700">{children}</h3>
        ),

        // 段落
        p: ({ children }) => (
          <p className="mb-3 text-gray-700 leading-relaxed">{children}</p>
        ),

        // 强调样式
        strong: ({ children }) => (
          <strong className="font-semibold text-gray-900">{children}</strong>
        ),
        em: ({ children }) => (
          <em className="italic text-gray-700">{children}</em>
        ),

        // 链接
        a: ({ href, children }) => (
          <a
            href={href}
            target="_blank"
            rel="noopener noreferrer"
            className="text-blue-600 hover:text-blue-700 underline"
          >
            {children}
          </a>
        ),

        // 列表
        ul: ({ children }) => (
          <ul className="list-disc list-inside mb-3 space-y-1">{children}</ul>
        ),
        ol: ({ children }) => (
          <ol className="list-decimal list-inside mb-3 space-y-1">{children}</ol>
        ),
        li: ({ children }) => (
          <li className="text-gray-700">
            <span className="ml-2">{children}</span>
          </li>
        ),

        // 引用块
        blockquote: ({ children }) => (
          <blockquote className="border-l-4 border-blue-500 pl-4 py-2 my-3 bg-blue-50 rounded-r">
            <div className="text-gray-700 italic">{children}</div>
          </blockquote>
        ),

        // 代码块
        code: ({ inline, className, children, ...props }: {inline?: boolean; className?: string; children: React.ReactNode}) => {
          const match = /language-(\w+)/.exec(className || '');
          const codeString = String(children).replace(/\n$/, '');
          const codeId = `code-${Math.random().toString(36).substr(2, 9)}`;

          if (!inline && match) {
            return (
              <div className="relative group my-3">
                <div className="absolute top-0 right-0 flex items-center space-x-2 p-2">
                  <span className="text-xs text-gray-500 bg-gray-100 px-2 py-1 rounded">
                    {match[1]}
                  </span>
                  <button
                    onClick={() => copyToClipboard(codeString, codeId)}
                    className="opacity-0 group-hover:opacity-100 transition-opacity p-1 hover:bg-gray-200 rounded"
                    title="复制代码"
                  >
                    {copiedCode === codeId ? (
                      <Check className="w-4 h-4 text-green-600" />
                    ) : (
                      <Copy className="w-4 h-4 text-gray-600" />
                    )}
                  </button>
                </div>
                <pre className="bg-gray-50 border border-gray-200 rounded-lg p-4 overflow-x-auto">
                  <code className={className} {...props}>
                    {children}
                  </code>
                </pre>
              </div>
            );
          }

          return (
            <code className="bg-gray-100 text-red-600 px-1.5 py-0.5 rounded text-sm" {...props}>
              {children}
            </code>
          );
        },

        // 表格
        table: ({ children }) => (
          <div className="overflow-x-auto my-4">
            <table className="min-w-full divide-y divide-gray-200 border border-gray-200 rounded-lg">
              {children}
            </table>
          </div>
        ),
        thead: ({ children }) => (
          <thead className="bg-gray-50">{children}</thead>
        ),
        tbody: ({ children }) => (
          <tbody className="bg-white divide-y divide-gray-200">{children}</tbody>
        ),
        tr: ({ children }) => (
          <tr className="hover:bg-gray-50">{children}</tr>
        ),
        th: ({ children }) => (
          <th className="px-4 py-2 text-left text-xs font-medium text-gray-500 uppercase tracking-wider">
            {children}
          </th>
        ),
        td: ({ children }) => (
          <td className="px-4 py-2 whitespace-nowrap text-sm text-gray-700">
            {children}
          </td>
        ),

        // 水平分割线
        hr: () => (
          <hr className="my-4 border-gray-200" />
        ),

        // 图片
        img: ({ src, alt }) => (
          <img
            src={src}
            alt={alt}
            className="max-w-full h-auto rounded-lg shadow-sm my-3"
          />
        ),
      }}
    >
      {content}
    </ReactMarkdown>
    </div>
  );
};

// ===== 自定义Alert组件 =====
interface AlertProps {
  type: 'info' | 'success' | 'warning' | 'error';
  children: React.ReactNode;
}

export const Alert: React.FC<AlertProps> = ({ type, children }) => {
  const configs = {
    info: {
      icon: <Info className="w-5 h-5" />,
      className: 'bg-blue-50 border-blue-200 text-blue-800',
      iconClass: 'text-blue-500'
    },
    success: {
      icon: <CheckCircle className="w-5 h-5" />,
      className: 'bg-green-50 border-green-200 text-green-800',
      iconClass: 'text-green-500'
    },
    warning: {
      icon: <AlertTriangle className="w-5 h-5" />,
      className: 'bg-yellow-50 border-yellow-200 text-yellow-800',
      iconClass: 'text-yellow-500'
    },
    error: {
      icon: <AlertCircle className="w-5 h-5" />,
      className: 'bg-red-50 border-red-200 text-red-800',
      iconClass: 'text-red-500'
    }
  };

  const config = configs[type];

  return (
    <div className={`flex items-start p-4 mb-3 border rounded-lg ${config.className}`}>
      <div className={`flex-shrink-0 ${config.iconClass}`}>
        {config.icon}
      </div>
      <div className="ml-3 flex-1">
        {children}
      </div>
    </div>
  );
};