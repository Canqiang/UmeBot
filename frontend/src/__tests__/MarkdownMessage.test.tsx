import React from 'react';
import { render, screen } from '@testing-library/react';
import { describe, it, expect } from 'vitest';
import { MarkdownMessage } from '../components/MarkdownMessage';

describe('MarkdownMessage', () => {
  const markdown = [
    '# Heading 1',
    '',
    '```javascript',
    'console.log(42);',
    '```',
    '',
    'Inline math $E=mc^2$ and block math:',
    '',
    '$$',
    'a^2 + b^2 = c^2',
    '$$',
    '',
  ].join('\n');

  it('renders markdown elements', () => {
    const { container } = render(<MarkdownMessage content={markdown} />);
    const heading = screen.getByRole('heading', { level: 1 });
    expect(heading.textContent).toBe('Heading 1');
    expect(container.querySelector('pre code')?.textContent).toBe('console.log(42);');
    expect(container.querySelector('.katex')).not.toBeNull();
    expect(heading).toMatchSnapshot();
  });
});
