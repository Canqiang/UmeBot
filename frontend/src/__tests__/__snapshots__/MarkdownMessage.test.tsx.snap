// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MarkdownMessage > renders markdown elements 1`] = `
<h1
  class="text-2xl font-bold mt-6 mb-4 text-gray-900"
>
  Heading 1
</h1>
`;
